lib/core/src/Cardano/Wallet/DB/Layer.hs

```diff
@@ -1395,194 +997,38 @@ deleteDelegationCertificates wid filters = do
--- and then concatenate the results afterwards. Used to avoid "too many SQL
--- variables" errors for "SELECT IN" queries.
-combineChunked :: [a] -> ([a] -> SqlPersistT IO [b]) -> SqlPersistT IO [b]
-combineChunked xs f = concatMapM f $ chunksOf chunkSize xs
 
+-- See also: issue #573.
 selectTxHistory
-    :: W.Wallet s
-    -> TimeInterpreter IO
+    :: Monad m
+    => W.Wallet s
+    -> TimeInterpreter m
     -> W.WalletId
     -> Maybe W.Coin
     -> W.SortOrder
-    -> [Filter TxMeta]
-    -> SqlPersistT IO [W.TransactionInfo]
-selectTxHistory cp ti wid minWithdrawal order conditions = do
-    let txMetaFilter = (TxMetaWalletId ==. wid):conditions
-    metas <- case minWithdrawal of
-        Nothing -> fmap entityVal <$> selectList txMetaFilter sortOpt
-        Just coin -> do
-            txids <- fmap (txWithdrawalTxId . entityVal)
-                <$> selectList [ TxWithdrawalAmount >=. coin ] []
-            ms <- combineChunked (nub txids) (\chunk -> selectList
-              ((TxMetaTxId <-. chunk):txMetaFilter) [])
-            let sortTxId = case order of
-                    W.Ascending -> sortOn (Down . txMetaTxId)
-                    W.Descending -> sortOn txMetaTxId
-            let sortSlot = case order of
-                    W.Ascending -> sortOn txMetaSlot
-                    W.Descending -> sortOn (Down . txMetaSlot)
-            pure $ sortSlot $ sortTxId $ fmap entityVal ms
-
-    let txids = map txMetaTxId metas
-    (ins, cins, outs, couts, ws) <- selectTxs txids
-
-    let tip = W.currentTip cp
-
-    liftIO $ txHistoryFromEntity
-        ti tip metas ins cins outs couts ws
-  where
-    -- Note: there are sorted indices on these columns.
-    -- The secondary sort by TxId is to make the ordering stable
-    -- so that testing with random data always works.
-    sortOpt = case order of
-        W.Ascending -> [Asc TxMetaSlot, Desc TxMetaTxId]
-        W.Descending -> [Desc TxMetaSlot, Asc TxMetaTxId]
-
-selectTxMeta
-    :: W.WalletId
-    -> W.Hash "Tx"
-    -> SqlPersistT IO (Maybe TxMeta)
-selectTxMeta wid tid =
-    fmap entityVal <$> selectFirst
-        [ TxMetaWalletId ==. wid, TxMetaTxId ==. (TxId tid)]
-        [ Desc TxMetaSlot ]
-
--- | Delete the transaction, but only if it's not in ledger.
--- Returns non-zero if this was a success.
-deletePendingOrExpiredTx
-    :: W.WalletId
-    -> W.Hash "Tx"
-    -> SqlPersistT IO Int
-deletePendingOrExpiredTx wid tid = do
-    let filt = [ TxMetaWalletId ==. wid, TxMetaTxId ==. (TxId tid) ]
-    selectFirst ((TxMetaStatus ==. W.InLedger):filt) [] >>= \case
-        Just _ -> pure 0  -- marked in ledger - refuse to delete
-        Nothing -> fromIntegral <$> deleteWhereCount
-            ((TxMetaStatus <-. [W.Pending, W.Expired]):filt)
+    -> (DB.TxMeta -> Bool)
+    -> TxWalletsHistory
+    -> m [W.TransactionInfo]
+selectTxHistory cp ti wid minWithdrawal order whichMeta
+    (decorateWithTxOuts ->  TxHistoryF txs, wmetas) = do
```

Comment:
I think that `ViewPatterns` is fairly confusing; I very much prefer `PatternSynonyms`. Suggestion:

```suggestion
    (txhistory, wmetas) = do
```

and

```
  TxHistoryF txs = decorateWithTxOuts txhistory
```

in a `where` clause.

Reply:
Everything we do not use is confusing at first!

Reply:
Pattern synonyms == the factored out where